opentelemetry/proto/trace/v1/trace.proto

```diff
@@ -249,6 +249,12 @@ message Span {
   // An optional final status for this span. Semantically when Status isn't set, it means
   // span's status code is unset, i.e. assume STATUS_CODE_UNSET (code = 0).
   Status status = 15;
+
+  // parent_span_is_remote identifies whether a span's parent is remote or not.
+  // This helps determine whether a span can be considered an entry-point span (also known as "local root span").
+  // A span is an entry-point span if it has no parent (parent_span_id is empty),
+  // or if parent_span_is_remote is true.
```

Comment:
What's the interoperability story here? All old OTLP sources that are unaware of this field will emit data such that parent_span_is_remote will arrive as false to all new OTLP receivers that are expecting the field. Is this what we want? What is the behavior of the receivers going to be when every single span that arrives has parent_span_is_remote=false?

Reply:
I also suggest to include a link to https://github.com/open-telemetry/oteps/blob/main/text/0182-otlp-remote-parent.md in the comments for readers who want to know more but are not familiar with the concept.

Reply:
Good point.
Since proto3 now supports optional, probably we should mark the field as such.

Reply:
Not all Protobuf generators generate presence functions for optional fields. IIRC, Gogoproto used by Collector doesn't. This means the data passed through the Collector will be destroyed (a present parent_span_is_remote=false will be forwarded with the field removed). Because of this in the past we avoided relying on optional fields. Things may have changed recently since I am not longer involved in the Collector actively.

@open-telemetry/collector-maintainers do we support optional fields in pdata currently?

Reply:
This is what the OTEP says BTW https://github.com/open-telemetry/oteps/blob/main/text/0182-otlp-remote-parent.md#internal-details:

> For backwards compatibility with older OTLP versions, the protobuf field should be nullable (true, false, or unspecified) and the opentelemetry-collector protogen code should provide an API that enables backend exporters to identify whether the field is set.

Reply:
I just checked how Collector supports the optionals and I think at least from Collector's perspective it was suboptimal to use them in OTLP:
- The Collector patches and replaces the optionals by oneof. (optional and oneof are wire compatible, which makes the patching possible).
- In memory oneof fields in Gogoproto are represented as an interface pointing to a struct. On 64bit arch that takes 16 bytes in the message struct plus an allocation for the struct containing the bool value (another 8 bytes on heap). This uses a ton of memory unnecessarily (at least 24 bytes) and has particularly bad performance due to the extra allocation.

Compare to enum:
- Same size as optional bool on the wire (for values of enum <127 and we only need 0..2). 
- In memory 4 bytes in the message struct (or at worst 8 bytes if int32 is padded to 64bits) and 0 allocations (so 3x or 6x less memory). Also much better performance because there is no extra allocation.

I am not familiar with how Protobuf Java code is generated. @Oberon00 you mentioned enum is an object. I wonder what the impact of that is on performance compared to optional boolean (which I imagine Protobuf would still need to model as something other than a primitive Boolean field in message's POD). If it is a wash between enum and optional in Java then I suggest we choose what is best for Go: stick to enums.

Reply:
I re-read the thread again and my recommendation is still to go with an enum. Have an "unspecified" value of 0 in the enum. If you need to differentiate between "false" and "unspecified" then make "false" another value. An obviously "true" needs to be another value of the enum.

One argument I see that can make me change my mind is if someone can demonstrate enums are very inefficient in languages other than Go (compared to optional boolean).

Reply:
@tigrannajaryan in java we manually serialize our data from a custom in memory representation (i.e. one not based off generated classes) to the protobuf wire format. Either is fine for us.

I'm fine with enum, but want to call out that this will create a precedent for using enums to represent a tri-states instead of optional booleans. That's fine with me but I think there are valid use cases for optional (such as the min / max fields) and I hope we continue to use them when appropriate.

Reply:
>  I think there are valid use cases for optional (such as the min / max fields) and I hope we continue to use them when appropriate.

If we already use optionals elsewhere, wouldn't that be an argument for using it here as well?

Reply:
The other places are floating point numbers, so not possible to represent them with enums. A boolean tri-state is unique because it can be represented in multiple ways.

Reply:
> I'm fine with enum, but want to call out that this will create a precedent for using enums to represent a tri-states instead of optional booleans.

Would be that bad? I'm wondering whether documenting this somewhere (pros and cons) would be enough?

Reply:
@tigrannajaryan We just discussed this in the Specification SIG and would like to get this PR in soon. What's your opinion on the type question? Are you fine with the above proposal of using `enum` in this case?

Reply:
It was @tigrannajaryan who suggested the enum, I think? IIUC, nobody objected against the enum, there was just a discussion if an optional boolean wouldn't be slightly better but that seems to be resolved now in favor of the enum.

Reply:
@estolfo Can you please change the PR to use an enum instead of "optional"? I think we have an agreement to move forward with the enum approach.

Reply:
Hi @Oberon00 I was also out for a few weeks. I'm back now and will make the change, thanks!

Reply:
I've defined an enum with
0 - unset
1 - false
2 - true

Is there a convention for representing tristate booleans as enums in protobuf definitions? Let me know if the enum should be adjusted.